### DISCUSSION TITLE
Forecasting without forecast rollout

### DISCUSSION THREAD

**Discussion by simonmoulds** [neuralhydrology/neuralhydrology#188] [Q&A] [answered]
First of all, thanks a lot for the work you do in developing and maintaining the package - it's really well designed and user-friendly. 

I have a question about using neuralhydrology to make forecasts, given that the library does not currently support forecast rollout. I'm using a HandoffForecastLSTM model to make seasonal forecasts. These are initialized monthly, and go out to around 90 days ahead in my application. During training, I replace forecast_inputs (let's say forecast precip and temperature) with the equivalent hindcast_inputs (observed precip and temperature). To make predictions, I create a pseudo basin for each initialization time (e.g. for basin id 1001, I create pseudo-basins 1001_20240101, 1001_20240201, ..., which each contain forecast data from the same initialization time, and which can be specified in `test_basin_file`). This approach does mean that the forecasts need to have similar statistical properties to the observations - i.e. they need bias correction (at least for seasonal climate forecasts - not sure about weather forecasts). 

I have a couple of questions: 

1. Given that the package doesn't currently support forecast rollout, do you think this approach is reasonable to make forecasts using the HandoffForecastLSTM model (and other forecast models), or is there a better way? I assume I could use a standard CudaLSTM model, but I like the flexibility of using additional inputs in the hindcast model. 
2. Supposing I wanted to add support for forecast rollout - from a cursory look through the code, I think this should be possible by creating a custom dataset class with a modified `__getitem__` method to ensure the hindcast and forecast periods align. Can you think of any other parts of the code that would require modification?

Many thanks!

**Comment by kratzert** [comment, marked as answer]
Hi @simonmoulds 

I was not really involved in the implementation of the hindcast/forecast stuff here in NH but here is what I would probably do.

Making edits to the basedataset can be quite tricky because this class has organically grown in some kind of a monster that even I don't like to touch anymore. What I would probably do is to write your own little dataset class all-together if you feel comfortable doing that. Note that our basedataset class include tons of features that 99.9% of the people don't use and most likely your own implementation would be much shorter and much more concise to what you want to do.

When we train forecasts models at Google, we usually structure our data in some an xr.Dataset first that might look like this:
![Screenshot from 2024-06-12 13-32-15](https://github.com/neuralhydrology/neuralhydrology/assets/13069767/8fbbf500-3afc-4826-8b32-e71af913810a)

Note that this dataset has a time index (datetime format), a lead_time index (integer, relative to the time index) and a basin index. And you see that some data variables are indexed only by basin and time (these are hindcast features) and others are indexed by basin, time, and lead_time (those are forecast features). Forecast features are aligned to the issue date of the forecast. So to get sample of hindcast + forecast features, we e.g. sample a basin and an issue time that for all hindcast features we `ds[hindcast_features].sel(basin=basin_id, time=slice(issue_time - seq_length + 1, issue_time + 1)` (note that the +1 is required since the end index is not included in the slice) and to get the aligned forecast features can be accessed by `ds[forecast_features].sel(basin=basin_id, time=issue_time)`.

Regarding the model itself: We used the HandoffForecastLSTM as our operational model two years ago but since then changed our architecture twice (there will be a blogpost about this sometime soon). The HandoffForecastLSTM has some annoying property in the forecast hairs that made it not suitable for our daily forecasts). Not sure how much impact this has on your task though.

Let me know if that helped you.

**Reply by simonmoulds** [reply]
Thanks @kratzert - that is helpful. Nice to see the structure of the data - that was along the lines of what I was thinking, which is encouraging. 

It's also helpful to know about the HandoffForecastLSTM. I was trying to reconcile the description in your recent Nature paper (for reference: _we developed a forecast model that uses an encoder–decoder model with one LSTM running over a historical sequence of meteorological (and geophysical) input data (the encoder LSTM) and another, separate, LSTM that runs over the 7-day forecast horizon with inputs from meteorological forecasts (the decoder LSTM)._) with the available forecast models and the HandoffForecastLSTM seemed closest, although the description suggests that you have removed the handoff network and pass the cell and hidden states from the encoder directly to the decoder... anyway, I look forward to the blog post! 

One last question - the approach you have outlined suggests that the forecast LSTM is trained directly on the forecasts. I can defintely see the advantages of this, but is there a danger that when the underlying met forecasts are wrong, the model is learning spurious relationships? E.g. when a flood peak occurs but the forecast has missed or underestimated the rainfall event. Perhaps at 5-days lead time the forecasts are sufficiently accurate that the impact of forecast inaccuracies is negligible?

Many thanks!
